avocado/core/nrunner.py

```diff
@@ -240,23 +298,108 @@ def task_from_recipe(task_path):
     runnable = Runnable(runnable_recipe.get('kind'),
                         runnable_recipe.get('uri'),
                         *runnable_recipe.get('args', ()))
-    return Task(identifier, runnable)
+    status_uris = recipe.get('status_uris')
+    return Task(identifier, runnable, status_uris)
+
+
+class StatusServer:
+
+    def __init__(self, uri, tasks_pending=None):
+        self.uri = uri
+        self.server_task = None
+        self.status = {}
+        if tasks_pending is None:
+            tasks_pending = []
+        self.tasks_pending = tasks_pending
+        self.wait_on_tasks_pending = len(self.tasks_pending) > 0
+
+    @asyncio.coroutine
+    def cb(self, reader, _):
+        while True:
+            if self.wait_on_tasks_pending:
+                if not self.tasks_pending:
+                    print('Status server: exiting due to all tasks finished')
+                    self.server_task.cancel()
+                    yield from self.server_task
+                    return True
+
+            message = yield from reader.readline()
+            if message == b'bye\n':
+                print('Status server: exiting due to user request')
+                self.server_task.cancel()
+                yield from self.server_task
+                return True
+
+            if not message:
+                return False
+
+            data = json_loads(message.strip())
+
+            if data['status'] not in ["init", "running"]:
+                try:
+                    self.tasks_pending.remove(data['id'])
+                    print('Task complete (%s): %s' % (data['status'],
+                                                      data['id']))
+                except IndexError:
+                    pass
+                except ValueError:
+                    pass
+                if data['status'] in self.status:
+                    self.status[data['status']] += 1
+                else:
+                    self.status[data['status']] = 1
+
+                if data['status'] not in ('pass', 'skip'):
```

Comment:
Just checking: `finished` is not in this set because you want its stdout/stderr printed if it a exec kind of runnable?